csds-client/client/util.go

```diff
@@ -0,0 +1,201 @@
+	envoy_config_filter_network_http_connection_manager_v2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	envoy_service_status_v2 "github.com/envoyproxy/go-control-plane/envoy/service/status/v2"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
+	"github.com/ghodss/yaml"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// helper method for parsing csds request yaml to nodematchers
+func ParseYaml(path string, yamlStr string, nms *[]*envoy_type_matcher.NodeMatcher) error {
+	if path != "" {
+		// parse yaml to json
+		filename, _ := filepath.Abs(path)
+		yamlFile, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+		js, err := yaml.YAMLToJSON(yamlFile)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		// parse the json array to a map to iterate it
+		var data map[string]interface{}
+		err = json.Unmarshal(js, &data)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		// parse each json object to proto
+		for _, n := range data["node_matchers"].([]interface{}) {
+			x := &envoy_type_matcher.NodeMatcher{}
+
+			jsonString, err := json.Marshal(n)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			err = protojson.Unmarshal(jsonString, x)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			*nms = append(*nms, x)
+		}
+	}
+	if yamlStr != "" {
+		var js []byte
+		var err error
+		// json input
+		if yamlStr[0] == '{' {
+			js = []byte(yamlStr)
+		} else {
+			// parse the yaml input into json
+			js, err = yaml.YAMLToJSON([]byte(yamlStr))
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+		}
+
+		// parse the json array to a map to iterate it
+		var data map[string]interface{}
+		err = json.Unmarshal(js, &data)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		// parse each json object to proto
+		i := 0
+		for _, n := range data["node_matchers"].([]interface{}) {
+			x := &envoy_type_matcher.NodeMatcher{}
+
+			jsonString, err := json.Marshal(n)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			err = protojson.Unmarshal(jsonString, x)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+
+			// merge the proto with existing proto from request_file
+			if i < len(*nms) {
+				proto.Merge((*nms)[i], x)
+			} else {
+				*nms = append(*nms, x)
+			}
+			i++
+		}
+	}
+	return nil
+}
+
+// parse gcp project number from metadata of nodematchers
+func ParseGCPProject(nms []*envoy_type_matcher.NodeMatcher) string {
+	for _, nm := range nms {
+		for _, mt := range nm.NodeMetadatas {
+			for _, path := range mt.Path {
+				if path.GetKey() == "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER" {
```

Comment:
Should this be set as a const in the module?

Does it have to have prefix "TRAFFICDIRECTOR" btw? It is just GCP project number, not necessarily specific to Traffic Director I think? In the future, is it possible that we might need to extract GCP project number to use some other endpoint?

Reply:
This is contract between TD and csds client. Right now it has trafficdirector to be consistent with what TD asks customer to put in node metadata to represent project number.

Reply:
This function is used to parse GCP project number for auto authentication mode on GCP. Since this client is designed to be generic, I'm not sure if this client can be used to connect to other uri besides TD for GCP. If it can, maybe I can modify this key based on the uri it connect to, since this key can change. If it will only connect to TD for GCP, maybe just make it const.

Reply:
To make this function truly generic, how about to name it something like

func getValueByKeyFromNodeMatcher(nms []*envoy_type_matcher.NodeMatcher, key string) string

Then introduce a const with TD GCP key name and call it where needed in the context of connecting to TD. This way we can keep it both generic, reusable and TD-specific.

Reply:
Make sense. Just changed it to `func getValueByKeyFromNodeMatcher(nms []*envoy_type_matcher.NodeMatcher, key string) string`. The key is specified before calling it based on the uri.